test/support/utils.rb

```diff
@@ -73,18 +74,32 @@ def create_pseudo_terminal(boot_options: "-r debug/run")
               when %r{Really quit\? \[Y/n\]}
                 cmd = @queue.pop
                 write.puts(cmd)
+              when /REPL ERROR/
+                error_messages = [line]
+
+                while line = read.gets
+                  case line.chomp
+                  # it means all error messages have been printed out
+                  when REPL_PROMPT
+                    break
+                  else
+                    error_messages << line
+                  end
+                end
+
+                raise "Debugger terminated because of: #{error_messages.join}"
```

Comment:
does it kill child process?

Reply:
yes.

```
# before
$ ps aux | grep ruby
st0012           13410   2.2  0.5  5833468 154264   ??  S     1:37下午   0:06.51 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           13412   1.1  0.4  5946096 144132   ??  S     1:37下午   0:04.57 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           13413   1.0  0.4  5421808 139928   ??  S     1:37下午   0:04.79 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           13409   0.6  0.1  5241596  40760   ??  S     1:37下午   0:02.77 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           14296   0.0  0.0  4399456    796 s000  S+    1:40下午   0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn ruby
```

```
$ ruby test/debug/watch_test.rb
Loaded suite test/debug/watch_test
Started
E
=========================================================================================================================================================================================================================
Error: test_debugger_only_stops_when_the_ivar_of_instance_changes(DEBUGGER__::ObjectInstanceVariableWatchingTest):
  RuntimeError: Debugger terminated because of: [REPL ERROR] #<RuntimeError: foo>
    /Users/st0012/projects/debug/lib/debug/session.rb:338:in `wait_command'
    /Users/st0012/projects/debug/lib/debug/session.rb:175:in `block (2 levels) in wait_command_loop'
    /Users/st0012/projects/debug/lib/debug/session.rb:174:in `loop'
    /Users/st0012/projects/debug/lib/debug/session.rb:174:in `block in wait_command_loop'
    /Users/st0012/projects/debug/lib/debug/session.rb:815:in `block in stop_all_threads'
    <internal:trace_point>:196:in `enable'
    /Users/st0012/projects/debug/lib/debug/session.rb:814:in `stop_all_threads'
    /Users/st0012/projects/debug/lib/debug/session.rb:173:in `wait_command_loop'
    /Users/st0012/projects/debug/lib/debug/session.rb:114:in `session_server_main'
    /Users/st0012/projects/debug/lib/debug/session.rb:73:in `block in initialize'
  INTERNAL_INFO: {"location":"/var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210616-14315-wlyagk.rb:6","line":6}

/Users/st0012/projects/debug/test/support/utils.rb:90:in `block (2 levels) in create_pseudo_terminal'
/Users/st0012/.rbenv/versions/2.7.2/lib/ruby/2.7.0/timeout.rb:95:in `block in timeout'
/Users/st0012/.rbenv/versions/2.7.2/lib/ruby/2.7.0/timeout.rb:33:in `block in kcatch'
/Users/st0012/.rbenv/versions/2.7.2/lib/ruby/2.7.0/timeout.rb:33:in `catch'
/Users/st0012/.rbenv/versions/2.7.2/lib/ruby/2.7.0/timeout.rb:33:in `catch'
/Users/st0012/.rbenv/versions/2.7.2/lib/ruby/2.7.0/timeout.rb:110:in `timeout'
/Users/st0012/projects/debug/test/support/utils.rb:59:in `block in create_pseudo_terminal'
/Users/st0012/projects/debug/test/support/utils.rb:55:in `spawn'
/Users/st0012/projects/debug/test/support/utils.rb:55:in `create_pseudo_terminal'
/Users/st0012/projects/debug/test/support/utils.rb:33:in `debug_code'
test/debug/watch_test.rb:57:in `test_debugger_only_stops_when_the_ivar_of_instance_changes'
     54:     end
     55:
     56:     def test_debugger_only_stops_when_the_ivar_of_instance_changes
  => 57:       debug_code(program) do
     58:         type 'continue'
     59:         assert_line_text('Student#initialize(name="John")')
     60:         type 'continue'
=========================================================================================================================================================================================================================

Finished in 0.226006 seconds.
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 0 assertions, 0 failures, 1 errors, 0 pendings, 0 omissions, 0 notifications
0% passed
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
4.42 tests/s, 0.00 assertions/s
```
```
# after
$ ps aux | grep ruby
st0012           13410   2.2  0.4  5817084 140212   ??  S     1:37下午   0:06.74 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           13412   1.2  0.4  5946096 146420   ??  S     1:37下午   0:04.67 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           13413   0.9  0.4  5421808 140436   ??  S     1:37下午   0:04.90 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           13409   0.6  0.1  5241596  40760   ??  S     1:37下午   0:02.84 /Users/st0012/.rbenv/versions/2.7.2/bin/ruby /Users/st0012/.rbenv/versions/2.7.2/bin/solargraph socket --port 0
st0012           14348   0.0  0.0  4399456    796 s000  S+    1:40下午   0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn ruby

```